### PR TITLE
[2d] fixes coming from clang-doxy

### PR DESCRIPTION
**Pull request by aPonza**
Fixup the warnings reported by doxygen built with clang parsing capabilities.

Notably, it includes a fix for the `Keypoint` class which was incorrectly not templated and instead interpreted as a namespace due to a lowercase initial in https://github.com/PointCloudLibrary/pcl/pull/3760#discussion_r402009423.

<details>
<summary>Before: list of warnings in module [2d]</summary>
<br>

```
error: /pcl/2d/include/pcl/2d/impl/convolution.hpp:45:1: error: variable has incomplete type 'void' [clang]
error: /pcl/2d/include/pcl/2d/impl/convolution.hpp:45:12: error: expected ';' at end of declaration [clang]
error: /pcl/2d/include/pcl/2d/impl/convolution.hpp:45:12: error: expected unqualified-id [clang]
error: /pcl/2d/include/pcl/2d/impl/edge.hpp:48:1: error: variable has incomplete type 'void' [clang]
error: /pcl/2d/include/pcl/2d/impl/edge.hpp:48:5: error: expected ';' at end of declaration [clang]
error: /pcl/2d/include/pcl/2d/impl/edge.hpp:48:5: error: expected unqualified-id [clang]
error: /pcl/2d/include/pcl/2d/impl/edge.hpp:85:26: error: qualified name refers into a specialization of variable template 'Edge' [clang]
error: /pcl/2d/include/pcl/2d/impl/kernel.hpp:45:1: error: variable has incomplete type 'void' [clang]
error: /pcl/2d/include/pcl/2d/impl/kernel.hpp:45:7: error: expected ';' at end of declaration [clang]
error: /pcl/2d/include/pcl/2d/impl/kernel.hpp:45:7: error: expected unqualified-id [clang]
error: /pcl/2d/include/pcl/2d/impl/kernel.hpp:95:15: error: qualified name refers into a specialization of variable template 'kernel' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:53:14: error: unknown type name 'ImageType' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:54:14: error: unknown type name 'ImageType' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:62:3: error: unknown type name 'ImageType' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:63:3: error: unknown type name 'ImageType' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:64:3: error: use of undeclared identifier 'conv_2d' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:65:3: error: use of undeclared identifier 'conv_2d' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:68:3: error: unknown type name 'ImageType' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:69:3: error: use of undeclared identifier 'conv_2d' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:72:3: error: unknown type name 'ImageType' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:73:3: error: use of undeclared identifier 'edge_detection' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:74:3: error: use of undeclared identifier 'edge_detection' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:74:44: error: use of undeclared identifier 'I_y' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:77:3: error: unknown type name 'ImageType' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:79:24: error: use of undeclared identifier 'I_y2' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:79:30: error: use of undeclared identifier 'I_y' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:79:35: error: use of undeclared identifier 'I_y' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:80:24: error: use of undeclared identifier 'I_xI_y' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:80:37: error: use of undeclared identifier 'I_y' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:83:3: error: unknown type name 'ImageType' [clang]
error: fatal error: too many errors emitted, stopping now [-ferror-limit=] [clang]
error: /pcl/2d/include/pcl/2d/impl/morphology.hpp:46:1: error: variable has incomplete type 'void' [clang]
error: /pcl/2d/include/pcl/2d/impl/morphology.hpp:46:11: error: expected ';' at end of declaration [clang]
error: /pcl/2d/include/pcl/2d/impl/morphology.hpp:46:11: error: expected unqualified-id [clang]
error: /pcl/2d/include/pcl/2d/impl/morphology.hpp:97:19: error: qualified name refers into a specialization of variable template 'Morphology' [clang]
error: /pcl/2d/include/pcl/2d/keypoint.h:50:3: error: use of class template 'Edge' requires template arguments [clang]
error: /pcl/2d/include/pcl/2d/keypoint.h:51:3: error: use of class template 'Convolution' requires template arguments [clang]
error: /pcl/2d/include/pcl/2d/keypoint.h:57:16: error: unknown type name 'ImageType' [clang]
error: /pcl/2d/include/pcl/2d/keypoint.h:58:16: error: unknown type name 'ImageType' [clang]
error: /pcl/2d/include/pcl/2d/keypoint.h:65:15: error: unknown type name 'ImageType' [clang]
error: /pcl/2d/include/pcl/2d/keypoint.h:65:34: error: unknown type name 'ImageType' [clang]
error: /pcl/2d/include/pcl/2d/keypoint.h:68:15: error: unknown type name 'ImageType' [clang]
error: /pcl/2d/include/pcl/2d/keypoint.h:69:15: error: unknown type name 'ImageType' [clang]
error: /pcl/2d/include/pcl/2d/keypoint.h:75:24: error: unknown type name 'ImageType' [clang]
error: /pcl/2d/include/pcl/2d/keypoint.h:75:43: error: unknown type name 'ImageType' [clang]
error: /pcl/2d/include/pcl/2d/keypoint.h:75:62: error: unknown type name 'ImageType' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:53:14: error: unknown type name 'ImageType' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:54:14: error: unknown type name 'ImageType' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:62:3: error: unknown type name 'ImageType' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:63:3: error: unknown type name 'ImageType' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:64:3: error: use of undeclared identifier 'conv_2d' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:65:3: error: use of undeclared identifier 'conv_2d' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:68:3: error: unknown type name 'ImageType' [clang]
error: /pcl/2d/include/pcl/2d/impl/keypoint.hpp:69:3: error: use of undeclared identifier 'conv_2d' [clang]
error: fatal error: too many errors emitted, stopping now [-ferror-limit=] [clang]
```

</details>

<details>
<summary>After: list of warnings in module [2d]</summary>
<br>
None
</details>

With the occasion, since I had to add 1 or 2 `#pragma once` directives, I also changed the header guards where possible and I will do this for all the subsequent PRs.

### CODE REVIEW

File: 2d/include/pcl/2d/impl/keypoint.hpp [hunk 1]
```diff
@@ -42,22 +42,23 @@
 #ifndef PCL_2D_KEYPOINT_HPP_
```
Comment (aPonza) [inline 2d/include/pcl/2d/impl/keypoint.hpp:42]: change to #pragma once

**Comment by aPonza** [comment]
You may want the top include to be the .h for a respective .hpp (similar to how you may want the .cpp to have its own .h at the top) and that should be enforceable with clang format as well (see `IncludeIsMainSourceRegex` [from 10 onwards](https://releases.llvm.org/10.0.0/tools/clang/docs/ReleaseNotes.html)) so I could touch up these files.

File: 2d/include/pcl/2d/keypoint.h [hunk 2]
```diff
@@ -45,10 +45,11 @@
 
 namespace pcl {
 
+template <typename ImageType>
```
[the remark this answers is not in the document]
Comment (aPonza) [reply 2d/include/pcl/2d/keypoint.h:48]: No clue, it was like this already, you can see it in the function signatures. Wanna change it up?

**Comment by aPonza** [comment]
BTW I also tried with
```
diff --git a/test/2d/keypoint_instantiation.cpp b/test/2d/keypoint_instantiation.cpp
index 6b8a915..e16db18 100644
--- a/test/2d/keypoint_instantiation.cpp
+++ b/test/2d/keypoint_instantiation.cpp
@@ -42,6 +42,12 @@
 /** This isn't useful except for testing the instantiation of this class. See #3898 */
 TEST (Keypoint, instantiatesWithAndWithoutPrecompiledHeaders)
 {
+#ifdef PCL_NO_PRECOMPILE
+  #pragma message "compiling without precompiled headers"
+#else // PCL_NO_PRECOMPILE
+  #pragma message "compiling with precompiled headers"
+#endif // PCL_NO_PRECOMPILE
+
   pcl::Keypoint<pcl::PointXYZ> keypoint = pcl::Keypoint<pcl::PointXYZ>();
   SUCCEED();
 }
```

and all seems fine.

File: test/2d/keypoint_instantiation.cpp [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+#include <pcl/2d/keypoint.h> // for pcl::Keypoint
+
+#include <pcl/point_types.h> // for pcl::PointXYZ
+#include <pcl/test/gtest.h>  // for SUCCEED
+
+/** This isn't useful except for testing the instantiation of this class. See #3898 */
+TEST (Keypoint, instantiatesWithAndWithoutPrecompiledHeaders)
+{
+  pcl::Keypoint<pcl::PointXYZ> keypoint = pcl::Keypoint<pcl::PointXYZ>();
```
Comment (kunaltyagi) [inline test/2d/keypoint_instantiation.cpp:45]: Which one is better:
* test (current)
* cpp file in 2d module (`grep` for `PCL_INSTANTIATE` and `PCL_NO_PRECOMPILE` for examples)

File: 2d/include/pcl/2d/keypoint.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (taketwo) [reply 2d/include/pcl/2d/keypoint.h:48]: Nevermind :)

File: test/2d/keypoint_instantiation.cpp (hunk 3 above)
Comment (taketwo) [reply test/2d/keypoint_instantiation.cpp:45]: `2d` module is header-only, I wouldn't change this just for the sake of instantiation.

**Pull request merged**